File: src/commands/say.js
import { PermissionFlagsBits } from 'discord.js';

export default {
  name: 'say',
  description: 'Faire dire un message au bot (prefix). Usage: !say [--webhook] message',
  async execute(message, args) {
    if (message.author.bot) return;

    const ownerId = process.env.OWNER_ID;
    const isOwner = ownerId && message.author.id === ownerId;
    const hasAdmin = message.member?.permissions?.has(PermissionFlagsBits.Administrator);
    if (!isOwner && !hasAdmin) {
      return message.reply('Vous n\'êtes pas autorisé à utiliser cette commande.');
    }

    let useWebhook = false;
    if (args[0] === '--webhook') {
      useWebhook = true;
      args.shift();
    }

    const content = args.join(' ').trim();
    if (!content) return message.reply('Veuillez fournir un message à envoyer.');

    try {
      if (useWebhook) {
        // Try to create a temporary webhook to better imitate an user
        const avatar = message.author.displayAvatarURL && message.author.displayAvatarURL();
        const webhook = await message.channel.createWebhook({ name: message.author.username, avatar });
        await webhook.send({ content });
        try { await webhook.delete('cleanup after say command'); } catch (e) { /* ignore */ }
      } else {
        await message.channel.send(content);
      }

      // Try to delete the invoking message to avoid leaving a trace
      try { await message.delete(); } catch (e) { /* ignore */ }
    } catch (err) {
      console.error('Erreur dans la commande say prefix:', err);
      return message.reply(`Erreur lors de l'envoi: ${err.message || String(err)}`);
    }
  }
};

File: Python/commands/say.py
"""Commande prefix `say` traduite depuis JS.
Usage: executed as `await execute(message, args)`
"""
import os
from logger import Logger
logger = Logger()


name = 'say'
description = "Faire dire un message au bot (prefix). Usage: !say [--webhook] message"


async def execute(message, args):
    try:
        if message.author.bot:
            return

        owner_id = os.getenv('OWNER_ID')
        is_owner = owner_id and str(message.author.id) == str(owner_id)
        has_admin = False
        try:
            has_admin = getattr(message.member.guild_permissions, 'administrator', False)
        except Exception:
            has_admin = False
        if not is_owner and not has_admin:
            return await message.reply("Vous n'êtes pas autorisé à utiliser cette commande.")

        use_webhook = False
        if args and len(args) > 0 and args[0] == '--webhook':
            use_webhook = True
            args = args[1:]

        content = ' '.join(args).strip()
        if not content:
            return await message.reply('Veuillez fournir un message à envoyer.')

        if use_webhook:
            try:
                # dynamic import of discord types
                webhook = None
                try:
                    webhook = await message.channel.create_webhook(name=str(message.author))
                except Exception:
                    webhook = None
                if webhook:
                    await webhook.send(content)
                    try:
                        await webhook.delete(reason='cleanup after say command')
                    except Exception:
                        logger.debug('webhook cleanup failed')
                    logger.info(f"say (prefix) sent via webhook by {getattr(message.author, 'name', message.author)}")
                else:
                    await message.channel.send(content)
                    logger.info('say sent as fallback message')
            except Exception as e:
                logger.error(f'Erreur say webhooks: {e}')
                return await message.reply(f"Erreur lors de l'envoi: {e}")
        else:
            try:
                await message.channel.send(content)
                logger.info('say (prefix) sent as bot message')
            except Exception as e:
                logger.error(f'Erreur say send: {e}')
                return await message.reply(f"Erreur lors de l'envoi: {e}")

        try:
            await message.delete()
        except Exception:
            logger.debug('invoking message delete failed')
        try:
            await message.channel.send("Ok, c'est envoyé.")
        except Exception:
            logger.debug('confirmation send failed')
    except Exception as err:
        logger.error(['Erreur dans la commande say prefix:', err])
        try:
            await message.reply(f"Erreur lors de l'envoi: {err}")
        except Exception:
            pass
